lasagne/layers/recurrent.py

```diff
@@ -0,0 +1,1239 @@
+        (p. 6) for further explanation.
+    unroll_scan : bool
+        If True the recursion is unrolled instead of using scan. For some
+        graphs this gives a significant speed up but it might also consume
+        more memory. When `unroll_scan` is true then the `gradient_steps`
+        setting is ignored. The input sequence length cannot be specified as
+        None when `unroll_scan` is True.
+    precompute_input : bool
+        If True, precompute input_to_hid before iterating through
+        the sequence. This can result in a speedup at the expense of
+        an increase in memory usage.
+
+    References
+    ----------
+    .. [1] Graves, Alex: "Generating sequences with recurrent neural networks."
+           arXiv preprint arXiv:1308.0850 (2013).
+    """
+    def __init__(self, incoming, input_to_hidden, hidden_to_hidden,
+                 nonlinearity=nonlinearities.rectify,
+                 hid_init=init.Constant(0.),
+                 backwards=False,
+                 learn_init=False,
+                 gradient_steps=-1,
+                 grad_clipping=False,
+                 unroll_scan=False,
+                 precompute_input=True,
+                 **kwargs):
+
+        super(CustomRecurrentLayer, self).__init__(incoming, **kwargs)
+
+        self.input_to_hidden = input_to_hidden
+        self.hidden_to_hidden = hidden_to_hidden
+        self.learn_init = learn_init
+        self.backwards = backwards
+        self.gradient_steps = gradient_steps
+        self.grad_clipping = grad_clipping
+        self.unroll_scan = unroll_scan
+        self.precompute_input = precompute_input
+
+        if unroll_scan and gradient_steps != -1:
+            raise ValueError(
+                "Gradient steps must be -1 when unroll_scan is true.")
+
+        if unroll_scan and self.input_shape[1] is None:
+            raise ValueError("Input sequence length cannot be specified as "
+                             "None when unroll_scan is True")
+
+        # Check that output shapes match
+        if input_to_hidden.output_shape != hidden_to_hidden.output_shape:
+            raise ValueError("The output shape for input_to_hidden and "
+                             "input_to_hidden must be equal, but "
+                             "input_to_hidden.output_shape={} and "
+                             "hidden_to_hidden.output_shape={}".format(
+                                 input_to_hidden.output_shape,
+                                 hidden_to_hidden.output_shape))
+
+        if nonlinearity is None:
+            self.nonlinearity = nonlinearities.identity
+        else:
+            self.nonlinearity = nonlinearity
+
+        # Get the input dimensionality and number of units based on the
+        # expected output of the input-to-hidden layer
+        self.num_inputs = np.prod(self.input_shape[2:])
+        self.num_units = input_to_hidden.output_shape[-1]
```

Comment:
I spent some time hacking on this yesterday and made some progress, but I agree that it can/should wait for another PR.  It'll be a backwards-compatible, but complicated-to-test change.